### PR TITLE
Update Kotlin and dokka to 1.7.0

### PR DESCRIPTION
**Pull request by famod**
I haven't checked yet whether [those Gradle tests are still failing on Java 19-EA](https://github.com/quarkusio/quarkus/pull/24086/files#diff-c4df82b52272e0d864fe167a9f60cedc3934c19527161ce00ff3aef2aaaf0387), but I don't think anything has changed in that regard.

### CODE REVIEW

File: jakarta/transform.sh [hunk 1]
```diff
@@ -34,7 +34,7 @@ if [ "${REWRITE_OFFLINE-false}" != "true" ]; then
   # Build Kotlin Maven Plugin to allow skipping main compilation
   # (skipping test compilation is supported but not main)
   rm -rf target/kotlin
-  git clone -b v1.6.21-jakarta --depth 1 https://github.com/gsmet/kotlin.git target/kotlin
+  git clone -b v1.7.0-jakarta --depth 1 https://github.com/gsmet/kotlin.git target/kotlin
```
Comment (famod) [inline jakarta/transform.sh:37]: @gsmet I suppose you'll have to do something to make this work?

**Comment by famod** [comment]
I have to take a closer look at those failures. That NPE was already reported via #26290.

File: jakarta/transform.sh (hunk 1 above)
Comment (gsmet) [reply jakarta/transform.sh:37]: Yes, I will have to. I need to create a PR for the Kotlin plugin. Maybe they will accept it.

**Comment by famod** [comment]
So it seems this needs some fundamental adjustments in `GradleApplicationModelBuilder` because `KotlinCompile` changed considerable in 1.7.0: https://github.com/quarkusio/quarkus/issues/26290#issuecomment-1166327208

I will not work on this because I'm not familiar all with that part of Quarkus/Gradle/Kotlin and my time is very limited.

@geoand @glefloch you are very welcome to take it from here.

**Comment by famod** [comment]
@evanchooly have you been able to look at Kotlin 1.7+? Can I close this PR?

**Comment by evanchooly** [comment]
I picked it back up again just today. We can close this one and I'll open a new one if you want.

**Comment by GavinRay97** [comment]
I think this should do it, but how do I test it?

```patch
diff --git a/devtools/gradle/gradle-model/build.gradle b/devtools/gradle/gradle-model/build.gradle
--- a/devtools/gradle/gradle-model/build.gradle	(revision 58cc7fb85891eb3b2f47c6a9a4db172393315b4f)
+++ b/devtools/gradle/gradle-model/build.gradle	(date 1657907839589)
@@ -1,5 +1,6 @@
 dependencies {
     implementation "io.quarkus:quarkus-core-deployment:${version}"
+    implementation "org.jetbrains.kotlin:kotlin-gradle-plugin-model:1.7.10"
     testImplementation "io.quarkus:quarkus-devtools-testing:${version}"
 }

diff --git a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java
--- a/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java	(revision 58cc7fb85891eb3b2f47c6a9a4db172393315b4f)
+++ b/devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/GradleApplicationModelBuilder.java	(date 1657908578186)
@@ -435,30 +436,21 @@
 
         final List<SourceDir> sourceDirs = new ArrayList<>(1);
         final List<SourceDir> resourceDirs = new ArrayList<>(1);
+
         project.getTasks().withType(AbstractCompile.class, t -> {
             if (!t.getEnabled()) {
                 return;
             }
             final FileTree source = t.getSource();
-            if (source.isEmpty()) {
-                return;
-            }
-            final File destDir = t.getDestinationDirectory().getAsFile().get();
-            if (!allClassesDirs.contains(destDir)) {
+            addCompileTaskFileTreeSourceDirs(allClassesDirs, sourceDirs, t, source);
+        });
+
+        project.getTasks().withType(KotlinJvmCompile.class, t -> {
+            if (!t.getEnabled()) {
                 return;
             }
-            final List<File> srcDirs = new ArrayList<>(1);
-            source.visit(a -> {
-                // we are looking for the root dirs containing sources
-                if (a.getRelativePath().getSegments().length == 1) {
-                    final File srcDir = a.getFile().getParentFile();
-                    if (srcDirs.add(srcDir)) {
-                        DefaultSourceDir sources = new DefaultSourceDir(srcDir.toPath(), destDir.toPath(),
-                                Collections.singletonMap("compiler", t.getName()));
-                        sourceDirs.add(sources);
-                    }
-                }
-            });
+            final FileTree source = t.getSources().getAsFileTree();
+            addCompileTaskFileTreeSourceDirs(allClassesDirs, sourceDirs, (AbstractCompile) t, source);
         });
 
         final File resourcesOutputDir = sourceSet.getOutput().getResourcesDir();
@@ -494,6 +486,28 @@
         module.addArtifactSources(new DefaultArtifactSources(classifier, sourceDirs, resourceDirs));
     }
 
+    private static void addCompileTaskFileTreeSourceDirs(FileCollection allClassesDirs, List<SourceDir> sourceDirs, AbstractCompile t, FileTree source) {
+        if (source.isEmpty()) {
+            return;
+        }
+        final File destDir = t.getDestinationDirectory().getAsFile().get();
+        if (!allClassesDirs.contains(destDir)) {
+            return;
+        }
+        final List<File> srcDirs = new ArrayList<>(1);
+        source.visit(a -> {
+            // we are looking for the root dirs containing sources
+            if (a.getRelativePath().getSegments().length == 1) {
+                final File srcDir = a.getFile().getParentFile();
+                if (srcDirs.add(srcDir)) {
+                    DefaultSourceDir sources = new DefaultSourceDir(srcDir.toPath(), destDir.toPath(),
+                            Collections.singletonMap("compiler", t.getName()));
+                    sourceDirs.add(sources);
+                }
+            }
+        });
+    }
+
     private void addSubstitutedProject(PathList.Builder paths, File projectFile) {
         File mainResourceDirectory = new File(projectFile, MAIN_RESOURCES_OUTPUT);
         if (mainResourceDirectory.exists()) {
```

**Pull request closed without merging**